File: JavaScript/Math-Array-String/13.js
// 1. Two Sum

/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var twoSum = function (nums, target) {
  const newMap = new Map();

  for (let i = 0; i < nums.length; i++) {
    const complement = target - nums[i];

    if (newMap.has(complement)) {
      return [newMap.get(complement), i];
    }

    newMap.set(nums[i], i);
  }
  return [];
};

console.log(twoSum([2, 5, 5, 11], 10));

// var twoSum = function(nums, target) {

//     for(let i=0;i<nums.length; i++){
//         for(let j= i+1; j<nums.length;j++){
//             if(nums[i] + nums[j] === target){
//                 return [i, j]
//             }
//         }
//     }
//     return []

// };

File: Python/Math-Array-String/13.py
# 1. Two Sum


from typing import List

class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        l = len(nums)
        for i in range(l - 1):
            for j in  range(i+1, l):
                if nums[i] + nums[j] == target:
                    return [i, j]
        return [] # No solution found
    

sol = Solution()
output = sol.twoSum([3, 3, 9], 12)
print(output)

#01
# class Solution:
#     def twoSum(self, nums: List[int], target: int) -> List[int]:
#         numMap = {}
#         l = len(nums)

#         for i in range(l):
#             numMap[nums[i]] = i

#         # Find the complement
#         for i in range(l):
#             complement = target - nums[i]
#             if complement in numMap and numMap[complement] != i:
#                 return [i, numMap[complement]]

#         return []  # No solution found


#02
# class Solution:
#     def twoSum(self, nums: List[int], target: int) -> List[int]:
#         numMap = {}
#         l = len(nums)

#         # Find the complement
#         for i in range(l):
#             complement = target - nums[i]
#             if complement in numMap and numMap[complement] != i:
#                 return [i, numMap[complement]]
#             numMap[nums[i]] = i

#         return []  # No solution found

